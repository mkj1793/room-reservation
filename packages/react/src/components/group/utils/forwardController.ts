import React, { FC } from 'react';

import { Controller, HTMLElementAttributesWithChildren, isFunctionOrObject } from '.';

const HDSForwardController = Symbol('HDSForwardController');

/**
 *
 * Basically cloned how React.ForwardRef works.
 * Sometimes it is convenient to pass controller to a React.JSX/FC component.
 * If a React.JSX/FC component has HDSForwardController as a prop,
 * then the controller should be passed as a second argument.
 *
 */

export function ForwardController<P extends HTMLElementAttributesWithChildren & Record<string, unknown>>(
  component: (props: HTMLElementAttributesWithChildren & P, controller: Controller) => React.ReactElement | null,
): FC<HTMLElementAttributesWithChildren & P> {
  const renderFunc = (componentProps: P & { forwardedController?: Controller }) => {
    const { forwardedController, ...rest } = componentProps;
    if (!forwardedController) {
      throw new Error('forwardedController not passed to ForwardController');
    }
    return component((rest as unknown) as P, forwardedController);
  };
  Object.defineProperty(renderFunc, HDSForwardController, {
    value: true,
  });
  return renderFunc;
}

export function isForwardController(component: unknown) {
  const reactObjectType = isFunctionOrObject(component) && Reflect.get(component as React.ComponentType, 'type');
  return isFunctionOrObject(reactObjectType) && Reflect.get(reactObjectType, HDSForwardController) === true;
}
