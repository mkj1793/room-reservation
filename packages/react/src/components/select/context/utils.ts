import React from 'react';

import { ChangeEvent } from './DataContext';

type DomHandlerProps = {
  id: string;
  type?: string;
  trigger: (event: ChangeEvent) => void;
};

export function createOnClickListener(props: DomHandlerProps) {
  const { id, type = 'click', trigger } = props;
  return {
    onClick: (originalEvent: React.MouseEvent) => {
      trigger({ id, type, payload: { originalEvent } });
    },
  };
}

export function createInputOnChangeListener(props: DomHandlerProps) {
  const { id, type = 'change', trigger } = props;
  return {
    onChange: (originalEvent: React.KeyboardEvent<HTMLInputElement>) => {
      trigger({
        id,
        type,
        payload: { value: originalEvent.currentTarget.value, originalEvent },
      });
    },
  };
}
