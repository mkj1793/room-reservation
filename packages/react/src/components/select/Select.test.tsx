import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { axe } from 'jest-axe';
import { isNull, isUndefined } from 'lodash';

import { Select } from './Select';
import { SearchResult, SelectProps } from './types';
import { getElementIds } from './utils';
import { createTimedPromise } from '../login/testUtils/timerTestUtil';
import { defaultId, defaultLabel, getSelectProps, renderWithHelpers } from './testUtil';

type ElementAttributes = {
  button: React.HTMLAttributes<HTMLButtonElement>;
  listAndInputContainer: React.HTMLAttributes<HTMLDivElement>;
  list: React.HTMLAttributes<HTMLDivElement> | React.HTMLAttributes<HTMLUListElement>;
  options: React.HTMLAttributes<HTMLLIElement> | React.HTMLAttributes<HTMLDivElement>;
  selectedOptions: React.HTMLAttributes<HTMLLIElement> | React.HTMLAttributes<HTMLDivElement>;
  listElementNodeName: string | undefined;
  group: React.HTMLAttributes<HTMLDivElement> | undefined;
  groupElementNodeName: string | undefined;
  groupLabel: React.HTMLAttributes<HTMLLIElement> | React.HTMLAttributes<HTMLDivElement> | undefined;
  input: React.HTMLAttributes<HTMLInputElement> | undefined;
};

type TestScenario = {
  description: string;
  selectOptions: SelectProps;
  expectedAttributes: ElementAttributes;
  groupCount: number;
  hasInput: boolean;
};
// type AxeViolation = Awaited<ReturnType<typeof axe>>['violations'][number];

describe('<Select />', () => {
  const convertToSameAttributeType = (target: unknown, comparedValue: unknown) => {
    const targetType = typeof target;
    if (targetType === 'boolean') {
      return comparedValue === 'true' || comparedValue === 'false' ? comparedValue === 'true' : comparedValue;
    }
    if (targetType === 'undefined') {
      return isNull(comparedValue) || isUndefined(comparedValue) ? undefined : comparedValue;
    }
    return String(comparedValue);
  };

  const getMismatchingElementAttributes = (
    el: HTMLElement,
    attributes: React.HTMLAttributes<HTMLElement> | undefined,
  ) => {
    const mismatches: string[] = [];
    if (!attributes) {
      return ['No attributes'];
    }
    Object.entries(attributes).forEach(([key, value]) => {
      // console.log('key', key, value, '---', el.getAttribute(key));
      const elementValue = convertToSameAttributeType(value, el.getAttribute(key));
      if (value !== elementValue) {
        mismatches.push(`${key} mismatch. Expected '${value}' but element has '${elementValue}'`);
      }
    });
    return mismatches;
  };

  describe('spec', () => {
    it('renders single select component with groups and filter input', () => {
      const { asFragment } = render(
        <Select {...getSelectProps({ groups: true, multiSelect: false, input: 'filter' })} />,
      );
      expect(asFragment()).toMatchSnapshot();
    });
    it('renders multi select component with groups and filter input', () => {
      const { asFragment } = render(
        <Select {...getSelectProps({ groups: true, multiSelect: true, input: 'filter' })} />,
      );
      expect(asFragment()).toMatchSnapshot();
    });
    it('single select component with groups and filter input should not have basic accessibility issues', async () => {
      const { container } = render(
        <Select {...getSelectProps({ groups: true, multiSelect: false, input: 'filter' })} />,
      );
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
    it('multi select component with groups and filter input should not have basic accessibility issues', async () => {
      const { container } = render(
        <Select {...getSelectProps({ groups: true, multiSelect: true, input: 'filter' })} />,
      );
      const results = await axe(container);

      expect(results).toHaveNoViolations();
    });
  });
  describe('List is opened', () => {
    it('list opens via button click', async () => {
      const { openList, getListItemLabels, options } = renderWithHelpers();
      await openList();
      const listItems = getListItemLabels();
      options.forEach((label, i) => {
        expect(listItems[i]).toBe(label);
      });
    });
  });

  describe('Accessibility attributes are set correctly', () => {
    const singleSelectNoGroupsNoInput: TestScenario = {
      description: 'Single select no groups no inputs',
      selectOptions: getSelectProps({ groups: false, multiSelect: false, input: undefined }),
      expectedAttributes: {
        button: {
          role: 'combobox',
          'aria-expanded': false,
          'aria-controls': getElementIds(defaultId).list,
          'aria-haspopup': 'listbox',
          'aria-activedescendant': undefined,
        },
        listAndInputContainer: {
          'aria-hidden': true,
          role: undefined,
          'aria-label': undefined,
        },
        list: {
          'aria-multiselectable': 'false',
          role: 'listbox',
        },
        options: {
          'aria-selected': 'false',
          role: 'option',
        },
        selectedOptions: {
          'aria-selected': 'true',
          role: 'option',
        },
        listElementNodeName: 'ul',
        groupElementNodeName: '',
        group: undefined,
        groupLabel: undefined,
        input: undefined,
      },
      groupCount: 0,
      hasInput: false,
    };

    const singleSelectWithGroupsNoInput: TestScenario = {
      description: 'Single select with groups no inputs',
      selectOptions: getSelectProps({ groups: true, multiSelect: false, input: undefined }),
      expectedAttributes: {
        button: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.button,
        },
        listAndInputContainer: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.listAndInputContainer,
          role: undefined,
          'aria-label': undefined,
        },
        list: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.list,
        },
        options: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.options,
        },
        selectedOptions: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.selectedOptions,
        },
        listElementNodeName: 'div',
        group: {
          role: 'group',
          'aria-label': 'SET_IN_CODE',
        },
        groupElementNodeName: 'ul',
        groupLabel: {
          role: 'presentation',
        },
        input: undefined,
      },
      groupCount: 3,
      hasInput: false,
    };

    const singleSelectNoGroupsWithInput: TestScenario = {
      description: 'Single select no groups with filter input',
      selectOptions: getSelectProps({ groups: false, multiSelect: false, input: 'filter' }),
      expectedAttributes: {
        button: {
          role: undefined,
          'aria-expanded': false,
          'aria-controls': getElementIds(defaultId).searchOrFilterInput,
          'aria-haspopup': 'dialog',
          'aria-activedescendant': undefined,
        },
        listAndInputContainer: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.listAndInputContainer,
          role: 'dialog',
          'aria-label': defaultLabel,
        },
        list: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.list,
        },
        options: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.options,
        },
        selectedOptions: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.selectedOptions,
        },
        listElementNodeName: 'ul',
        groupElementNodeName: '',
        group: undefined,
        groupLabel: undefined,
        input: {
          role: 'combobox',
          'aria-controls': getElementIds(defaultId).list,
          'aria-activedescendant': '', // this should change!
          'aria-haspopup': 'listbox',
        },
      },
      groupCount: 0,
      hasInput: true,
    };

    const multiSelectNoGroupsNoInput: TestScenario = {
      description: 'Multiselect no groups no inputs',
      selectOptions: getSelectProps({ groups: false, multiSelect: true, input: undefined }),
      expectedAttributes: {
        button: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.button,
        },
        listAndInputContainer: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.listAndInputContainer,
        },
        list: {
          'aria-multiselectable': 'true',
          role: 'listbox',
          'aria-label': undefined,
        },
        options: {
          'aria-selected': 'false',
          role: 'option',
        },
        selectedOptions: {
          'aria-selected': 'true',
          role: 'option',
        },
        listElementNodeName: 'ul',
        groupElementNodeName: '',
        group: undefined,
        groupLabel: undefined,
        input: undefined,
      },
      groupCount: 0,
      hasInput: false,
    };

    const multiSelectWithGroupsNoInput: TestScenario = {
      description: 'Multiselect with groups no inputs',
      selectOptions: getSelectProps({ groups: true, multiSelect: true, input: undefined }),
      expectedAttributes: {
        button: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.button,
          'aria-haspopup': 'dialog',
        },
        listAndInputContainer: {
          'aria-hidden': true,
          role: 'dialog',
          'aria-label': `${defaultLabel}. 12 choices.`,
        },
        list: {
          // these undefined values indicate that the element should not have special attributes
          'aria-multiselectable': undefined,
          'aria-label': undefined,
          role: undefined,
        },
        options: {
          'aria-checked': 'false',
          role: 'checkbox',
        },
        selectedOptions: {
          'aria-checked': 'true',
          role: 'checkbox',
        },
        listElementNodeName: 'div',
        group: {
          role: 'group',
          'aria-label': undefined,
        },
        groupElementNodeName: 'div',
        groupLabel: {
          role: 'checkbox',
          'aria-checked': false,
        },
        input: undefined,
      },
      groupCount: 3,
      hasInput: false,
    };

    const singleSelectWithGroupsFilterInput: TestScenario = {
      description: 'Single select with groups with filter input',
      selectOptions: getSelectProps({ groups: true, multiSelect: false, input: 'filter' }),
      expectedAttributes: {
        button: {
          role: undefined,
          'aria-expanded': false,
          'aria-controls': getElementIds(defaultId).searchOrFilterInput,
          'aria-haspopup': 'dialog',
          'aria-activedescendant': undefined,
        },
        listAndInputContainer: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.listAndInputContainer,
          role: 'dialog',
          'aria-label': defaultLabel,
        },
        list: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.list,
        },
        options: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.options,
        },
        selectedOptions: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.selectedOptions,
        },
        listElementNodeName: singleSelectWithGroupsNoInput.expectedAttributes.listElementNodeName,
        groupElementNodeName: singleSelectWithGroupsNoInput.expectedAttributes.groupElementNodeName,
        group: singleSelectWithGroupsNoInput.expectedAttributes.group,
        groupLabel: singleSelectWithGroupsNoInput.expectedAttributes.groupLabel,
        input: {
          role: 'combobox',
          'aria-controls': getElementIds(defaultId).list,
          'aria-activedescendant': '', // this should change!
          'aria-haspopup': 'listbox',
        },
      },
      groupCount: 3,
      hasInput: true,
    };

    const multiSelectWithGroupsFilterInput: TestScenario = {
      description: 'MultiSelect with groups with filter input',
      selectOptions: getSelectProps({ groups: true, multiSelect: true, input: 'filter' }),
      expectedAttributes: {
        button: {
          role: undefined,
          'aria-expanded': false,
          'aria-controls': getElementIds(defaultId).searchOrFilterInput,
          'aria-haspopup': 'dialog',
          'aria-activedescendant': undefined,
        },
        listAndInputContainer: {
          ...singleSelectNoGroupsNoInput.expectedAttributes.listAndInputContainer,
          role: 'dialog',
          'aria-label': defaultLabel,
        },
        list: {
          'aria-multiselectable': undefined,
          'aria-label': `12 choices.`,
          role: 'dialog',
        },
        options: {
          ...multiSelectWithGroupsNoInput.expectedAttributes.options,
        },
        selectedOptions: {
          ...multiSelectWithGroupsNoInput.expectedAttributes.selectedOptions,
        },
        listElementNodeName: 'div',
        groupElementNodeName: 'div',
        group: {
          role: 'group',
          'aria-label': undefined,
        },
        groupLabel: {
          role: 'checkbox',
          'aria-checked': false,
        },
        input: {
          role: 'combobox',
          'aria-controls': getElementIds(defaultId).list,
          'aria-activedescendant': '', // this should change!
          'aria-haspopup': 'dialog',
        },
      },
      groupCount: 3,
      hasInput: true,
    };
    const multiSelectNoGroupsFilterInput: TestScenario = {
      description: 'MultiSelect no groups with filter input',
      selectOptions: getSelectProps({ groups: false, multiSelect: true, input: 'filter' }),
      expectedAttributes: {
        button: {
          role: undefined,
          'aria-expanded': false,
          'aria-controls': getElementIds(defaultId).searchOrFilterInput,
          'aria-haspopup': 'dialog',
          'aria-activedescendant': undefined,
        },
        listAndInputContainer: {
          ...multiSelectWithGroupsFilterInput.expectedAttributes.listAndInputContainer,
        },
        list: {
          ...multiSelectNoGroupsNoInput.expectedAttributes.list,
        },
        options: {
          ...multiSelectNoGroupsNoInput.expectedAttributes.options,
        },
        selectedOptions: {
          ...multiSelectNoGroupsNoInput.expectedAttributes.selectedOptions,
        },
        listElementNodeName: 'ul',
        groupElementNodeName: undefined,
        group: undefined,
        groupLabel: undefined,
        input: {
          ...singleSelectNoGroupsWithInput.expectedAttributes.input,
        },
      },
      groupCount: 0,
      hasInput: true,
    };
    [
      singleSelectNoGroupsNoInput,
      singleSelectWithGroupsNoInput,
      singleSelectNoGroupsWithInput,
      singleSelectWithGroupsFilterInput,
      multiSelectNoGroupsNoInput,
      multiSelectWithGroupsNoInput,
      multiSelectNoGroupsFilterInput,
      multiSelectWithGroupsFilterInput,
    ].forEach(({ selectOptions, description, expectedAttributes, groupCount, hasInput }) => {
      describe(`${description}`, () => {
        it('Button has correct attributes before and after opened', async () => {
          const { openList, getButtonElement, getAllListElements } = renderWithHelpers(selectOptions);
          expect(getMismatchingElementAttributes(getButtonElement(), expectedAttributes.button)).toHaveLength(0);
          await openList();
          expect(
            getMismatchingElementAttributes(getButtonElement(), {
              ...expectedAttributes.button,
              'aria-expanded': true,
              'aria-activedescendant': hasInput ? undefined : String(getAllListElements()[0].getAttribute('id')),
            }),
          ).toHaveLength(0);
        });
        it('List and input container has correct attributes', async () => {
          const { openList, getListAndInputContainer } = renderWithHelpers(selectOptions);
          expect(
            getMismatchingElementAttributes(getListAndInputContainer(), expectedAttributes.listAndInputContainer),
          ).toHaveLength(0);
          await openList();
          expect(
            getMismatchingElementAttributes(getListAndInputContainer(), {
              ...expectedAttributes.listAndInputContainer,
              'aria-hidden': false,
            }),
          ).toHaveLength(0);
        });
        it('List element has correct attributes and node', async () => {
          const { getListElement } = renderWithHelpers(selectOptions);
          const listElement = getListElement();
          // no need to open it.
          expect(getMismatchingElementAttributes(listElement, expectedAttributes.list)).toHaveLength(0);
          expect(listElement.nodeName.toLowerCase()).toBe(expectedAttributes.listElementNodeName);
        });
        it('Groups are created inside the list and have correct attributes', async () => {
          const { getGroupElements, openList, groupsAndOptions } = renderWithHelpers(selectOptions);
          expect(getGroupElements()).toHaveLength(0);
          if (!groupCount) {
            return;
          }
          await openList();
          const groupElements = getGroupElements();
          expect(groupElements).toHaveLength(groupCount);
          groupElements.forEach((groupElement, index) => {
            const { label } = groupsAndOptions[index];
            expect(getMismatchingElementAttributes(groupElement, { 'aria-label': label })).toHaveLength(0);
            expect(groupElement.nodeName.toLowerCase()).toBe(expectedAttributes.groupElementNodeName);
          });
        });
        it('Group labels have correct attributes', async () => {
          const { getGroupLabelElements, openList } = renderWithHelpers(selectOptions);
          expect(getGroupLabelElements()).toHaveLength(0);
          if (!groupCount) {
            return;
          }
          await openList();
          const groupLabelElements = getGroupLabelElements();
          expect(groupLabelElements).toHaveLength(groupCount);
          groupLabelElements.forEach((groupElement) => {
            expect(getMismatchingElementAttributes(groupElement, expectedAttributes.groupLabel)).toHaveLength(0);
          });
        });
        it('Input element has correct attributes', async () => {
          const { getInputElement, openList } = renderWithHelpers(selectOptions);
          if (!hasInput) {
            return;
          }
          await openList();
          expect(getMismatchingElementAttributes(getInputElement(), expectedAttributes.input)).toHaveLength(0);
        });
        it('Options have correct attributes before and after selections', async () => {
          const {
            openList,
            getOptionElements,
            isListOpen,
            clickOptionAndWaitForRerender,
            getGroupLabelElements,
            filterSelectedOptions,
            isElementSelected,
          } = renderWithHelpers(selectOptions);
          await openList();

          getOptionElements().forEach((optionElement) => {
            expect(getMismatchingElementAttributes(optionElement, expectedAttributes.options)).toHaveLength(0);
            expect(isElementSelected(optionElement)).toBeFalsy();
          });
          // do not click #0 option or multiselect group is selected
          const indexOfClickedOption = 1;
          expect(filterSelectedOptions(getOptionElements())).toHaveLength(0);
          await clickOptionAndWaitForRerender(indexOfClickedOption);
          expect(isListOpen()).toBe(!!selectOptions.multiSelect);
          await openList();
          const groupLabels = getGroupLabelElements();
          getOptionElements().forEach((optionElement, index) => {
            const expectedOptionAttributes =
              index === indexOfClickedOption
                ? { ...expectedAttributes.selectedOptions }
                : { ...expectedAttributes.options };
            const isMultiSelectGroupLabel =
              selectOptions.multiSelect && groupCount > 0 && groupLabels.includes(optionElement);
            const isGroupLabelOfSelectedOption = isMultiSelectGroupLabel && index < indexOfClickedOption;
            if (isGroupLabelOfSelectedOption) {
              expectedOptionAttributes['aria-checked'] = 'mixed';
            }
            expect(getMismatchingElementAttributes(optionElement, expectedOptionAttributes)).toHaveLength(0);
            expect(isElementSelected(optionElement)).toBe(index === indexOfClickedOption);
          });
        });
      });
    });
  });
  describe('Filter', () => {
    it('Filter hides options and clearing filtering restores options', async () => {
      const { openList, getOptionElements, setInputValue } = renderWithHelpers(
        getSelectProps({ groups: false, multiSelect: false, input: 'filter' }),
      );
      await openList();
      expect(getOptionElements()).toHaveLength(3);
      await setInputValue('Option 1');
      await waitFor(() => {
        expect(getOptionElements()).toHaveLength(1);
      });
      await setInputValue('');
      await waitFor(() => {
        expect(getOptionElements()).toHaveLength(3);
      });
    });
    it('Screen reader notifications are rendered', async () => {
      const { openList, setInputValue, getScreenReaderNotifications } = renderWithHelpers(
        getSelectProps({ groups: false, multiSelect: false, input: 'filter' }),
      );
      await openList();
      await setInputValue('Option 1');
      await waitFor(
        () => {
          expect(getScreenReaderNotifications()).toHaveLength(1);
        },
        { interval: 500 },
      );
      const notification = getScreenReaderNotifications()[0] as string;
      expect(notification.includes('Option 1')).toBeTruthy();
      expect(notification.includes('Found 1 option')).toBeTruthy();
    });
  });
  describe('Search', () => {
    it('Search updates all data', async () => {
      const resultArray = ['Result 1', 'Result 2'];
      const onSearch: SelectProps['onSearch'] = () => {
        return createTimedPromise({ options: resultArray }, 300) as Promise<SearchResult>;
      };
      const { openList, getOptionElements, setInputValue, getListItemLabels } = renderWithHelpers({
        ...getSelectProps({ groups: false, multiSelect: false, input: undefined }),
        onSearch,
      });
      await openList();
      expect(getOptionElements()).toHaveLength(3);
      await setInputValue('Option 1');
      await waitFor(() => {
        expect(getListItemLabels()).toEqual(resultArray);
      });
      await setInputValue('');
      await waitFor(() => {
        expect(getOptionElements()).toHaveLength(0);
      });
    });
    it('User is notified if there are no results', async () => {
      const { openList, setInputValue, getSearchAndFilterInfoTexts } = renderWithHelpers({
        ...getSelectProps({ groups: false, multiSelect: false, input: undefined }),
        onSearch: () => createTimedPromise({ options: [] }, 300) as Promise<SearchResult>,
      });
      await openList();
      await setInputValue('Option 1');
      await waitFor(
        () => {
          expect(getSearchAndFilterInfoTexts()).toHaveLength(2);
        },
        { interval: 500 },
      );
      const notification = getSearchAndFilterInfoTexts()[0] as string;
      expect(notification.includes(`No options found for "Option 1"`)).toBeTruthy();
    });
  });
});
