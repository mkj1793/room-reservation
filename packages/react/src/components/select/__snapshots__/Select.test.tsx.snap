// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> spec renders multi select component with groups and filter input 1`] = `
<DocumentFragment>
  <div
    class="wrapper root"
    id="select-component"
    tabindex="-1"
  >
    <label
      class="label "
      for="select-component-dropdown-button"
      id="select-component-label"
    >
      Label
    </label>
    <div
      class="selectAndListContainer"
    >
      <div
        class="selectedOptionsContainer"
      >
        <button
          aria-controls="select-component-input-element"
          aria-expanded="false"
          aria-haspopup="dialog"
          aria-invalid="false"
          aria-label="Label. . 0 options selected."
          class="dropdownButton selectedOptions placeholder spaceForOneDigit"
          id="select-component-dropdown-button"
          type="button"
        >
          <span>
            <svg
              aria-hidden="true"
              aria-label="location"
              class="icon s"
              role="img"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M11.9669 1.5C14.0265 1.5 16.0869 2.27765 17.6579 3.83367C20.8001 6.945 20.5878 11.7938 17.6579 15.1017L17.0345 15.8107C14.4235 18.8021 12.9688 20.7706 11.9669 22.7477C10.8937 20.6185 9.28527 18.4993 6.27797 15.1017C3.34814 11.7938 3.13508 6.945 6.27797 3.83367C7.84906 2.27765 9.90869 1.5 11.9669 1.5ZM11.999 3.5C10.3743 3.5 8.84685 4.12598 7.69956 5.26179C6.56645 6.38341 5.97897 7.81769 6.00058 9.41066C6.02151 10.984 6.64884 12.5453 7.76709 13.8087L8.42773 14.5606C10.0167 16.3842 11.1442 17.8 12.0004 19.0644C12.9748 17.6274 14.2916 16.0006 16.233 13.8087C17.3512 12.5453 17.9786 10.984 17.9995 9.41066C18.0198 7.81769 17.4323 6.38341 16.3005 5.26179C15.1519 4.12598 13.6251 3.5 11.999 3.5ZM12.0004 6C13.9338 6 15.5 7.56702 15.5 9.50039C15.5 11.433 13.9338 13 12.0004 13C10.067 13 8.5 11.433 8.5 9.50039C8.5 7.56702 10.067 6 12.0004 6ZM12.0003 8C11.1725 8 10.5 8.67319 10.5 9.50034C10.5 10.3268 11.1725 11 12.0003 11C12.8275 11 13.5 10.3268 13.5 9.50034C13.5 8.67319 12.8275 8 12.0003 8Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          <div
            aria-hidden="true"
            class="labels"
          >
            <span
              class="dropdownButtonOption"
            />
          </div>
        </button>
        <button
          aria-hidden="true"
          class="dropdownButton icon arrowButton"
          id="select-component-arrow-button"
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            aria-label="angle-down"
            class="icon s angleIcon"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12 13.5L17 8.5L18.5 10L12 16.5L5.5 10L7 8.5L12 13.5Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        aria-hidden="true"
        aria-label="Label"
        class="listAndInputContainer withSearchOrFilter"
        role="dialog"
      >
        <div
          class="root searchOrFilterInput"
        >
          <label
            class="label "
            for="select-component-input-element"
            id="select-component-input-element-label"
          >
            Filter
          </label>
          <div
            class="inputWrapper"
          >
            <input
              aria-activedescendant=""
              aria-controls="select-component-list"
              aria-expanded="true"
              aria-haspopup="dialog"
              aria-labelledby="select-component-input-element-label select-component-label"
              class="input hasClearButton"
              id="select-component-input-element"
              placeholder="Type text to filter results with"
              role="combobox"
              type="text"
              value=""
            />
            <div
              class="buttonWrapper"
            >
              <button
                aria-label="Clear filter"
                class="button clearButton"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  aria-label="cross-circle"
                  class="icon s"
                  role="img"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2ZM12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4ZM15 7.5L16.5 9L13.5 12L16.5 15L15 16.5L12 13.5L9 16.5L7.5 15L10.5 12L7.5 9L9 7.5L12 10.5L15 7.5Z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="listContainer hidden"
          style="max-height: 286px;"
        >
          <div
            aria-labelledby="select-component-choices-count"
            class="list shiftOptions multiSelectList"
            id="select-component-list"
            role="dialog"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
    <div
      aria-live="polite"
      class="screenReaderNotifications"
    />
  </div>
</DocumentFragment>
`;

exports[`<Select /> spec renders single select component with groups and filter input 1`] = `
<DocumentFragment>
  <div
    class="wrapper root"
    id="select-component"
    tabindex="-1"
  >
    <label
      class="label "
      for="select-component-dropdown-button"
      id="select-component-label"
    >
      Label
    </label>
    <div
      class="selectAndListContainer"
    >
      <div
        class="selectedOptionsContainer"
      >
        <button
          aria-controls="select-component-input-element"
          aria-expanded="false"
          aria-haspopup="dialog"
          aria-invalid="false"
          aria-label="Label. . 0 options selected."
          class="dropdownButton selectedOptions placeholder spaceForOneDigit"
          id="select-component-dropdown-button"
          type="button"
        >
          <span>
            <svg
              aria-hidden="true"
              aria-label="location"
              class="icon s"
              role="img"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M11.9669 1.5C14.0265 1.5 16.0869 2.27765 17.6579 3.83367C20.8001 6.945 20.5878 11.7938 17.6579 15.1017L17.0345 15.8107C14.4235 18.8021 12.9688 20.7706 11.9669 22.7477C10.8937 20.6185 9.28527 18.4993 6.27797 15.1017C3.34814 11.7938 3.13508 6.945 6.27797 3.83367C7.84906 2.27765 9.90869 1.5 11.9669 1.5ZM11.999 3.5C10.3743 3.5 8.84685 4.12598 7.69956 5.26179C6.56645 6.38341 5.97897 7.81769 6.00058 9.41066C6.02151 10.984 6.64884 12.5453 7.76709 13.8087L8.42773 14.5606C10.0167 16.3842 11.1442 17.8 12.0004 19.0644C12.9748 17.6274 14.2916 16.0006 16.233 13.8087C17.3512 12.5453 17.9786 10.984 17.9995 9.41066C18.0198 7.81769 17.4323 6.38341 16.3005 5.26179C15.1519 4.12598 13.6251 3.5 11.999 3.5ZM12.0004 6C13.9338 6 15.5 7.56702 15.5 9.50039C15.5 11.433 13.9338 13 12.0004 13C10.067 13 8.5 11.433 8.5 9.50039C8.5 7.56702 10.067 6 12.0004 6ZM12.0003 8C11.1725 8 10.5 8.67319 10.5 9.50034C10.5 10.3268 11.1725 11 12.0003 11C12.8275 11 13.5 10.3268 13.5 9.50034C13.5 8.67319 12.8275 8 12.0003 8Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          <div
            aria-hidden="true"
            class="labels"
          >
            <span
              class="dropdownButtonOption"
            />
          </div>
        </button>
        <button
          aria-hidden="true"
          class="dropdownButton icon arrowButton"
          id="select-component-arrow-button"
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            aria-label="angle-down"
            class="icon s angleIcon"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12 13.5L17 8.5L18.5 10L12 16.5L5.5 10L7 8.5L12 13.5Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        aria-hidden="true"
        aria-label="Label"
        class="listAndInputContainer withSearchOrFilter"
        role="dialog"
      >
        <div
          class="root searchOrFilterInput"
        >
          <label
            class="label "
            for="select-component-input-element"
            id="select-component-input-element-label"
          >
            Filter
          </label>
          <div
            class="inputWrapper"
          >
            <input
              aria-activedescendant=""
              aria-controls="select-component-list"
              aria-expanded="true"
              aria-haspopup="listbox"
              aria-labelledby="select-component-input-element-label select-component-label"
              class="input hasClearButton"
              id="select-component-input-element"
              placeholder="Type text to filter results with"
              role="combobox"
              type="text"
              value=""
            />
            <div
              class="buttonWrapper"
            >
              <button
                aria-label="Clear filter"
                class="button clearButton"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  aria-label="cross-circle"
                  class="icon s"
                  role="img"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2ZM12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4ZM15 7.5L16.5 9L13.5 12L16.5 15L15 16.5L12 13.5L9 16.5L7.5 15L10.5 12L7.5 9L9 7.5L12 10.5L15 7.5Z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="listContainer hidden"
          style="max-height: 286px;"
        >
          <div
            aria-multiselectable="false"
            class="list"
            id="select-component-list"
            role="listbox"
            tabindex="-1"
          />
        </div>
      </div>
    </div>
    <div
      aria-live="polite"
      class="screenReaderNotifications"
    />
  </div>
</DocumentFragment>
`;
